File: internal/provideradapters/claude.go
package provideradapters

import (
	"strings"
	"time"

	"github.com/g960059/agtmux/internal/stateengine"
)

type ClaudeAdapter struct{}

func NewClaudeAdapter() *ClaudeAdapter {
	return &ClaudeAdapter{}
}

func (a *ClaudeAdapter) ID() string {
	return stateengine.ProviderClaude
}

func (a *ClaudeAdapter) DetectProvider(meta stateengine.PaneMeta) (float64, bool) {
	return detectByAgentOrCmd(meta, stateengine.ProviderClaude, "claude", "claude-code", "cc")
}

func (a *ClaudeAdapter) BuildEvidence(meta stateengine.PaneMeta, now time.Time) []stateengine.Evidence {
	combined := normalizeForMatch(meta.RawReasonCode, meta.LastEventType, meta.SessionLabel, meta.PaneTitle)
	source := strings.ToLower(strings.TrimSpace(meta.StateSource))
	kind := kindFromSource(source)
	evidence := make([]stateengine.Evidence, 0, 4)

	if hasAnyToken(combined, "approval", "waiting_approval", "needs_approval", "permission") {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityWaitingApproval, kind, source, "claude:approval", 0.98, 0.96))
	}
	if hasAnyToken(combined, "waiting_input", "input_required", "await_user", "prompt") {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityWaitingInput, kind, source, "claude:input", 0.92, 0.9))
	}
	if hasAnyToken(combined, "error", "failed", "panic", "exception") {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityError, kind, source, "claude:error", 1.0, 0.95))
	}
	runningHint := hasAnyToken(combined,
		"working",
		"running",
		"in_progress",
		"streaming",
		"task_started",
		"agent_turn_started",
		"pretooluse",
	)
	if runningHint {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityRunning, kind, source, "claude:running_signal", 0.9, 0.86))
	}
	if hasAnyToken(combined, "idle", "completed", "done", "stop", "wrapper_exit", "session_end") {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityIdle, kind, source, "claude:idle_signal", 0.88, 0.88))
	}
	// Claude false-positive suppression: poller-running without explicit running hints should prefer idle.
	if strings.EqualFold(strings.TrimSpace(meta.RawState), stateengine.ActivityRunning) &&
		source == "poller" &&
		!runningHint {
		evidence = append(evidence, buildEvidence(now, a.ID(), stateengine.ActivityIdle, stateengine.EvidenceCapture, source, "claude:poller_running_without_signal", 0.93, 0.9))
	}
	return evidence
}

File: crates/agtmux-core/src/adapt/providers/claude.rs
use crate::adapt::{detect_by_agent_or_cmd, EvidenceBuilder, ProviderDetector};
use crate::types::{ActivityState, Evidence, EvidenceKind, PaneMeta, Provider, SourceType};
use chrono::{DateTime, Utc};
use std::time::Duration;

pub struct ClaudeDetector;

impl ProviderDetector for ClaudeDetector {
    fn id(&self) -> Provider {
        Provider::Claude
    }

    fn detect(&self, meta: &PaneMeta) -> Option<f64> {
        detect_by_agent_or_cmd(meta, "claude", &["claude", "claude-code", "cc"])
    }
}

pub struct ClaudeEvidenceBuilder;

impl EvidenceBuilder for ClaudeEvidenceBuilder {
    fn provider(&self) -> Provider {
        Provider::Claude
    }

    fn build_evidence(&self, meta: &PaneMeta, now: DateTime<Utc>) -> Vec<Evidence> {
        let combined = format!(
            "{} {} {}",
            meta.raw_state, meta.raw_reason_code, meta.last_event_type
        )
        .to_lowercase();

        let mut evidence = Vec::new();

        let has = |tokens: &[&str]| tokens.iter().any(|t| combined.contains(t));

        if has(&["approval", "waiting_approval", "needs_approval", "permission"]) {
            evidence.push(Evidence {
                provider: Provider::Claude,
                kind: EvidenceKind::PollerMatch("approval".into()),
                signal: ActivityState::WaitingApproval,
                weight: 0.95,
                confidence: 0.92,
                timestamp: now,
                ttl: Duration::from_secs(180),
                source: SourceType::Poller,
                reason_code: meta.raw_reason_code.clone(),
            });
        }

        if has(&["waiting_input", "input_required", "await_user", "prompt"]) {
            evidence.push(Evidence {
                provider: Provider::Claude,
                kind: EvidenceKind::PollerMatch("waiting_input".into()),
                signal: ActivityState::WaitingInput,
                weight: 0.90,
                confidence: 0.88,
                timestamp: now,
                ttl: Duration::from_secs(180),
                source: SourceType::Poller,
                reason_code: meta.raw_reason_code.clone(),
            });
        }

        if has(&["error", "failed", "panic", "exception"]) {
            evidence.push(Evidence {
                provider: Provider::Claude,
                kind: EvidenceKind::PollerMatch("error".into()),
                signal: ActivityState::Error,
                weight: 1.00,
                confidence: 0.95,
                timestamp: now,
                ttl: Duration::from_secs(180),
                source: SourceType::Poller,
                reason_code: meta.raw_reason_code.clone(),
            });
        }

        if has(&[
            "working",
            "running",
            "in_progress",
            "streaming",
            "task_started",
            "agent_turn_started",
            "pretooluse",
        ]) {
            evidence.push(Evidence {
                provider: Provider::Claude,
                kind: EvidenceKind::PollerMatch("running".into()),
                signal: ActivityState::Running,
                weight: 0.90,
                confidence: 0.86,
                timestamp: now,
                ttl: Duration::from_secs(90),
                source: SourceType::Poller,
                reason_code: meta.raw_reason_code.clone(),
            });
        }

        if has(&["idle", "completed", "done", "stop", "session_end"]) {
            evidence.push(Evidence {
                provider: Provider::Claude,
                kind: EvidenceKind::PollerMatch("idle".into()),
                signal: ActivityState::Idle,
                weight: 0.88,
                confidence: 0.88,
                timestamp: now,
                ttl: Duration::from_secs(90),
                source: SourceType::Poller,
                reason_code: meta.raw_reason_code.clone(),
            });
        }

        evidence
    }
}
